### PR TITLE
feat(contrib): add eureka registry (#1792)

### PR DESCRIPTION
**Pull request by weetime**
- feat(examples): add eureka registry example (#1792)

#### Description (what this PR does / why we need it):

Many Java service with SpringCloud use eureka.
We hope to achive registry interface to Kratos framework to access middleware eureka

#### Which issue(s) this PR fixes (resolves / be part of):

resolves  (#1792)

### CODE REVIEW

**Comment by weetime** [comment]
I want to append  
./contrib/registry/eureka 
to /hack/.test_ignored_files.

Test Eureka registry dependent on EurekaServer like consul.

**Comment by haiyux** [comment]
把go.mod 的地址改成`github.com/go-kratos/kratos/contrib/registry/eureka/v2`呀 要不然请求不到的

File: contrib/registry/eureka/go.mod [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module git.haodf.net/go/rocky/contrib/registry/eureka/v2
```
Comment (Cluas) [inline contrib/registry/eureka/go.mod:1]: git.haodf.net?

Comment (weetime) [reply contrib/registry/eureka/go.mod]: fix done.

File: contrib/registry/eureka/client.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	FetchAllUpInstances(ctx context.Context) []Instance
+	FetchAppInstances(ctx context.Context, appID string) (m Application, err error)
+	FetchAppUpInstances(ctx context.Context, appID string) []Instance
+	FetchAppInstance(ctx context.Context, appID string, instanceID string) (m Instance, err error)
+	FetchInstance(ctx context.Context, instanceID string) (m Instance, err error)
+	Out(ctx context.Context, appID, instanceID string) error
+	Down(ctx context.Context, appID, instanceID string) error
+}
+
+type Opt func(e *Client)
```
Comment (Cluas) [inline contrib/registry/eureka/client.go:122]: It is better to unify the code style，https://github.com/go-kratos/kratos/blob/9662ef3c213d3f7802c27d6b85386bd649a76de7/contrib/registry/etcd/registry.go#L19 for etcd, https://github.com/go-kratos/kratos/blob/9662ef3c213d3f7802c27d6b85386bd649a76de7/contrib/registry/nacos/registry.go#L30 for nacos, https://github.com/go-kratos/kratos/blob/9662ef3c213d3f7802c27d6b85386bd649a76de7/contrib/registry/consul/registry.go#L21 for consul, etc...

File: contrib/registry/eureka/client.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+)
+
+const (
+	statusUp            = "UP"
+	statusDown          = "DOWN"
+	statusOutOfServeice = "OUT_OF_SERVICE"
+	heartbeatRetry      = 3
+)
+
+const tpl = `{
```
Comment (Cluas) [inline contrib/registry/eureka/client.go:25]: I'm confused, isn't this just plain json? What is the benefit of using template.

File: contrib/registry/eureka/client.go (hunk 2 above)
Comment (weetime) [reply contrib/registry/eureka/client.go:122]: I will refactor it.

File: contrib/registry/eureka/client.go (hunk 3 above)
Comment (weetime) [reply contrib/registry/eureka/client.go]: remove done.

**Review by Cluas — DISMISSED** [review]

File: contrib/registry/eureka/client.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+			e.heartbeatInterval = duration
+		}
+	}
+}
+
+func WithCtx(ctx context.Context) ClientOption {
+	return func(e *Client) { e.ctx = ctx }
+}
+
+func WithPath(path string) ClientOption {
```
Comment (tonybase) [inline contrib/registry/eureka/client.go:119]: WithNamespace

File: contrib/registry/eureka/client.go [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+func WithHeartbeatInterval(interval string) ClientOption {
+	return func(e *Client) {
+		if duration := e.toDuration(interval); duration > 0 {
+			e.heartbeatInterval = duration
+		}
+	}
+}
+
+func WithCtx(ctx context.Context) ClientOption {
```
Comment (tonybase) [inline contrib/registry/eureka/client.go:115]: WithContext()

File: contrib/registry/eureka/client.go [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+	Down(ctx context.Context, appID, instanceID string) error
+}
+
+type ClientOption func(e *Client)
+
+func WithMaxRetry(maxRetry int) ClientOption {
+	return func(e *Client) { e.maxRetry = maxRetry }
+}
+
+func WithHeartbeatInterval(interval string) ClientOption {
```
Comment (tonybase) [inline contrib/registry/eureka/client.go:107]: WithHeartbeatInterval(d time.Duration)

File: contrib/registry/eureka/client.go [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+
+func NewClient(urls []string, opts ...ClientOption) *Client {
+	tr := &http.Transport{
+		MaxIdleConns: maxIdleConns,
+	}
+
+	e := &Client{
+		ctx:               context.Background(),
+		urls:              urls,
+		eurekaPath:        "eureka/v2",
```
Comment (tonybase) [inline contrib/registry/eureka/client.go:142]: /microservices

File: contrib/registry/eureka/client.go (hunk 5 above)
Comment (weetime) [reply contrib/registry/eureka/client.go:115]: This package has the same nam with WithContext.
Rename WithClientContext.

File: contrib/registry/eureka/client.go (hunk 4 above)
Comment (weetime) [reply contrib/registry/eureka/client.go:119]: I will refactor it.

File: contrib/registry/eureka/client.go (hunk 6 above)
Comment (weetime) [reply contrib/registry/eureka/client.go:107]: With string we can use 1m replace 60.
I will refactor to ensure consistency with this project.

File: contrib/registry/eureka/client.go (hunk 7 above)
Comment (weetime) [reply contrib/registry/eureka/client.go:142]: This is eureka http path,old version use 'eureka',new version with 'eureka/v2'
https://github.com/Netflix/eureka/wiki/Eureka-REST-operations

**Review by tonybase — APPROVED** [review]

**Comment by weetime** [comment]
./contrib/registry/discovery) (pull_request) Fail after merge main.
What's happened

**Comment by tonybase** [comment]
@weetime You can merge main branch to resolve Lint failures.

**Review by Cluas — APPROVED** [review]

**Comment by tonybase** [comment]
> ./contrib/registry/discovery) (pull_request) Fail after merge main. What's happened

```
return ins, nil
```

you can remove the returns value to fix lint problem

**Pull request merged**